src/pyk/proof/exploration.py

```diff
@@ -0,0 +1,347 @@
+from __future__ import annotations
+
+import json
+import logging
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from pyk.kore.rpc import LogEntry
+
+from ..kast.manip import flatten_label
+from ..kcfg import KCFG
+from ..prelude.ml import mlAnd, mlTop
+from ..utils import ensure_dir_path, hash_str, shorten_hashes
+from .equality import ProofSummary
+from .proof import CompositeSummary, Proof, ProofStatus, Prover
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable, Mapping
+    from pathlib import Path
+    from typing import Any, Final, TypeVar
+
+    from ..kast.inner import KInner
+    from ..kast.outer import KClaim
+    from ..kcfg import KCFGExplore
+    from ..kcfg.kcfg import NodeIdLike
+
+    T = TypeVar('T', bound='Proof')
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+class ExplorationProof(Proof):
+    kcfg: KCFG
+    init: int
+    logs: dict[int, tuple[LogEntry, ...]]
+
+    _terminal_node_ids: set[int]
+
+    def __init__(
+        self,
+        id: str,
+        kcfg: KCFG,
+        init: NodeIdLike,
+        logs: dict[int, tuple[LogEntry, ...]],
+        proof_dir: Path | None = None,
+        terminal_node_ids: Iterable[NodeIdLike] | None = None,
+        subproof_ids: Iterable[str] = (),
+        admitted: bool = False,
+    ):
+        if type(self) == ExplorationProof:
+            if not subproof_ids:
+                raise ValueError('Subproofs provided for an ExplorationProof')
+            if not admitted:
+                raise ValueError('Admitted ExplorationProof')
```

Comment:
Should it not be

```suggestion
        if type(self) == ExplorationProof:
            if subproof_ids:
                raise ValueError('Subproofs provided for an ExplorationProof')
            if admitted:
                raise ValueError('Admitted ExplorationProof')
```

instead?

Reply:
Ways for eliminating this special casing, as per our Slack discussion:
* Pull `admitted` and `subproofs` from `Proof` into subclasses that use them
* Make `ExplorationProof` abstract so that `APRProof` can implement it. Make basic implementation for `ExplorationProver` to instantiate.
* `APRProver` is not an `ExplorationProver`, just uses one. `APRPRoof` is no longer an `ExplorationProof`.

Reply:
Yes yes, that was bizarre.

Reply:
> Should it not be
> 
> instead?

Yes yes, that was bizarre.